activesupport/lib/active_support/rescuable.rb

```diff
@@ -65,65 +63,132 @@ def rescue_from(*klasses, &block)
           elsif klass.is_a?(String)
             klass
           else
-            raise ArgumentError, "#{klass} is neither an Exception nor a String"
+            raise ArgumentError, "#{klass.inspect} must be an Exception class or a String referencing an Exception class"
           end
 
           # Put the new handler at the end because the list is read in reverse.
-          self.rescue_handlers += [[key, options[:with]]]
+          self.rescue_handlers += [[key, with]]
         end
       end
-    end
 
-    # Tries to rescue the exception by looking up and calling a registered handler.
-    def rescue_with_handler(exception)
-      if handler = handler_for_rescue(exception)
-        handler.arity != 0 ? handler.call(exception) : handler.call
-        true # don't rely on the return value of the handler
+      # Handles an exception by looking up and calling a registered handler.
+      # Does not raise the exception if there's no registered handler. Use
+      #
+      #     rescue => exception
+      #       rescue_with_handler(exception) || raise
+      #
+      # to re-raise unhandled exceptions.
+      #
+      # By default, this only looks for handlers matching the exception
+      # itself. It can check for handlers matching the exception's cause,
+      # too, by passing <tt>and_cause: true</tt>.
+      #
+      # When an exception is raised in response to another one, such as
+      # raising a generic outer error to indicate one of many internal
+      # errors, the outer exception's cause is set to the original exception:
+      #
+      #     begin
+      #       raise original
+      #     rescue
+      #       raise wrapped
+      #     end
+      #
+      # Since we raised wrapped while handling original,
+      #
+      #     wrapped.cause == original
+      #
+      # When an exception with a #cause is handled, we use the highest
+      # priority handler for both the exception and its cause. That means
+      # that a handler declared for one exception class may be called with
+      # an exception of a different class (the cause):
+      #
+      #     rescue_from WrappedException, with: …
+      #     rescue_from OriginalException, with: …
+      #
+      # The last handler declaration is highest priority, so raising a
+      # wrapped exception with an original exception as its cause will be
+      # handled by the OriginalException handler, not the WrappedException
+      # handler.
+      #
+      # The aim here is to provide generic exception handling that can be
+      # overridden with handling for more specific causes, like a generic
+      # Active Record error handling and a specific PostgreSQL unique
+      # constraint violation error.
```

Comment:
I have doubts about this resolution order... it seems too easy to end up handling a cause in place of the outer exception.

"Wrapper is generic, cause is specific" is an accurate description of AR's exception structure, but I think that's a (regrettable) special case, not the norm. More often, I'd expect cause to be something like an `Errno::*`, whereas the wrapper describes what's gone wrong *in application terms*. The "generic" angle is supposed to be addressed by the exception's inheritance hierarchy.

If I've rescued an exception and reraised it as MySpecificUserErrorException, I definitely don't expect a later `rescue_from ActiveRecord::StatementInvalid` to hijack that ahead of my designated handler: I wrapped it specifically to hide that implementation detail.

So, I think I'm arguing we should handle the wrapped exception if there is any handler available for it, and only fall back to the cause as a last resort. Conveniently, that also generalizes better for `cause.cause` etc. If you know you have a particular exception class that implements the "generic wrapper" pattern, you can write a handler for it which eagerly tries to handle the cause first... but I really think that's a property of the exception class, not of the caller. (Read: with this change, I'm not sure we'd need `and_cause` to be optional.)